File: templates/java_rest_api_template/src/main/java/com/template/restapi/service/ExampleItemService.java
/**
 * #############################################################################
 * ### Example service
 * ###
 * ### @file ExampleItemService.java
 * ### @author Sebastian Russo
 * ### @date 2026
 * #############################################################################
 *
 * Business-logic layer between controllers and the repository. Controllers stay
 * thin (HTTP concerns only) while the service owns rules, validation calls,
 * id generation, and mapping to response DTOs.
 *
 */
package com.template.restapi.service;

import java.util.List;
import java.util.UUID;

import com.template.restapi.dto.ExampleItemCreate;
import com.template.restapi.dto.ExampleItemResponse;
import com.template.restapi.dto.ExampleItemUpdate;
import com.template.restapi.entity.ExampleItem;
import com.template.restapi.error.ResourceNotFoundException;
import com.template.restapi.repository.ExampleItemRepository;
import com.template.restapi.resources.cache.RedisCacheService;
import com.template.restapi.util.CustomLogger;

import org.springframework.stereotype.Service;

@Service
public class ExampleItemService {

    private static final String CACHE_KEY_PREFIX = "example_item:";
    private static final int CACHE_TTL_SECONDS = 600;

    private final ExampleItemRepository repository;
    private final RedisCacheService cacheService;

    public ExampleItemService(ExampleItemRepository repository, RedisCacheService cacheService) {
        this.repository = repository;
        this.cacheService = cacheService;
    }

    public List<ExampleItemResponse> listAll() {
        CustomLogger.debug("Listing all example items");
        return repository.findAll().stream()
                .map(ExampleItemResponse::from)
                .toList();
    }

    public ExampleItemResponse getById(String id) {
        String cacheKey = CACHE_KEY_PREFIX + id;

        var cached = cacheService.cacheGet(cacheKey, ExampleItemResponse.class);
        if (cached.isPresent()) {
            CustomLogger.debug("Cache hit for example item " + id);
            return cached.get();
        }

        ExampleItem item = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "Example item with id '" + id + "' not found."));
        ExampleItemResponse response = ExampleItemResponse.from(item);
        cacheService.cacheSet(cacheKey, response, CACHE_TTL_SECONDS);
        CustomLogger.debug("Cache miss for example item " + id);
        return response;
    }

    public ExampleItemResponse create(ExampleItemCreate body) {
        String id = UUID.randomUUID().toString();
        ExampleItem item = new ExampleItem(id, body.name(), body.description());
        repository.save(item);
        CustomLogger.info("Created example item with id '" + id + "'");
        return ExampleItemResponse.from(item);
    }

    public ExampleItemResponse update(String id, ExampleItemUpdate body) {
        ExampleItem item = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "Example item with id '" + id + "' not found."));

        // Partial update: only apply the fields that were actually provided.
        if (body.name() != null) {
            item.setName(body.name());
        }
        if (body.description() != null) {
            item.setDescription(body.description());
        }
        repository.save(item);
        cacheService.cacheDelete(CACHE_KEY_PREFIX + id);
        CustomLogger.info("Updated example item with id '" + id + "'");
        return ExampleItemResponse.from(item);
    }

    public void delete(String id) {
        if (!repository.deleteById(id)) {
            throw new ResourceNotFoundException("Example item with id '" + id + "' not found.");
        }
        cacheService.cacheDelete(CACHE_KEY_PREFIX + id);
        CustomLogger.info("Deleted example item with id '" + id + "'");
    }
}

File: templates/csharp_rest_api_template/src/RestApiTemplate/Services/ExampleItemService.cs
using Template.RestApi.DTOs;
using Template.RestApi.Entities;
using Template.RestApi.Errors;
using Template.RestApi.Repositories;
using Template.RestApi.Resources.Cache;
using Template.RestApi.Utils;

namespace Template.RestApi.Services;

public sealed class ExampleItemService(IExampleItemRepository repository, RedisCacheService cacheService)
{
    private const string CacheKeyPrefix = "example_item:";
    private const int CacheTtlSeconds = 600;

    public IReadOnlyList<ExampleItemResponse> ListAll() => repository.FindAll().Select(ExampleItemResponse.From).ToList();

    public ExampleItemResponse GetById(string id)
    {
        var cacheKey = CacheKeyPrefix + id;
        var cached = cacheService.CacheGet<ExampleItemResponse>(cacheKey);
        if (cached is not null)
        {
            CustomLogger.Debug($"Cache hit for example item {id}");
            return cached;
        }

        var response = ExampleItemResponse.From(Find(id));
        cacheService.CacheSet(cacheKey, response, CacheTtlSeconds);
        CustomLogger.Debug($"Cache miss for example item {id}");
        return response;
    }

    public ExampleItemResponse Create(ExampleItemCreate input)
    {
        var item = new ExampleItem(Guid.NewGuid().ToString(), input.Name, input.Description);
        repository.Save(item);
        CustomLogger.Info($"Created example item with id '{item.Id}'.");
        return ExampleItemResponse.From(item);
    }

    public ExampleItemResponse Update(string id, ExampleItemUpdate input)
    {
        var item = Find(id);
        if (input.Name is not null) item.Name = input.Name;
        if (input.Description is not null) item.Description = input.Description;
        repository.Save(item);
        cacheService.CacheDelete(CacheKeyPrefix + id);
        return ExampleItemResponse.From(item);
    }

    public void Delete(string id)
    {
        if (!repository.DeleteById(id))
        {
            throw new ResourceNotFoundException($"Example item with id '{id}' was not found.");
        }

        cacheService.CacheDelete(CacheKeyPrefix + id);
    }

    private ExampleItem Find(string id) => repository.FindById(id) ?? throw new ResourceNotFoundException($"Example item with id '{id}' was not found.");
}
